language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/extensions/ballerina/symbol/BallerinaSymbolServiceImpl.java

```diff
@@ -41,4 +68,78 @@
+            }
+
+            try {
+                expressionTypeResponse.setDocumentIdentifier(new TextDocumentIdentifier(fileUri));
+
+                // Get the semantic model.
+                Optional<SemanticModel> semanticModel = this.workspaceManager.semanticModel(filePath.get());
+
+                if (semanticModel.isPresent()) {
+                    Optional<Symbol> symbol = Optional.empty();
+                    for (int i = 0; i < request.getPosition().offset(); i++) {
+                        Document document = workspaceManager.document(filePath.get()).get();
+                        symbol = semanticModel.get().symbol(document, LinePosition.from(request.getPosition().line(),
+                                request.getPosition().offset() - i));
+                        if (symbol.isPresent()) {
+                            break;
+                        }
+                    }
+                    if (symbol.isPresent()) {
+                        expressionTypeResponse.setTypes(getType(symbol.get()));
+                        return expressionTypeResponse;
+                    }
+                }
+                return expressionTypeResponse;
+            } catch (Throwable e) {
+                String msg = "Operation 'ballerinaSymbol/type' failed!";
+                this.clientLogger.logError(LSContextOperation.DOC_TYPE, msg, e, request.getDocumentIdentifier(),
+                        (Position) null);
+                return expressionTypeResponse;
+            }
+        });
+    }
+
+    private List<String> getType(Symbol symbol) {
+        List<String> allTypes = new ArrayList<>();
+        if (symbol.kind() == SymbolKind.VARIABLE) {
+            VariableSymbol variableSymbol = (VariableSymbol) symbol;
+            if (variableSymbol.typeDescriptor().typeKind() == TypeDescKind.UNION) {
+                BallerinaUnionTypeSymbol unionTypeSymbol = (BallerinaUnionTypeSymbol) variableSymbol.typeDescriptor();
+                for (TypeSymbol typeSymbol: unionTypeSymbol.memberTypeDescriptors()) {
+                    if (!allTypes.contains(typeSymbol.typeKind().getName())) {
+                        allTypes.add(typeSymbol.typeKind().getName());
+                    }
+                }
+            } else if (variableSymbol.typeDescriptor().typeKind() == TypeDescKind.TYPE_REFERENCE) {
+                allTypes.add(variableSymbol.typeDescriptor().getName().get());
+            } else {
+                allTypes.add(variableSymbol.typeDescriptor().typeKind().getName());
+            }
+        } else if (symbol.kind() == SymbolKind.METHOD) {
+            MethodSymbol methodSymbol = (MethodSymbol) symbol;
+            TypeSymbol returnTypeSymbol = methodSymbol.typeDescriptor().returnTypeDescriptor().get();
+            if (returnTypeSymbol.typeKind() == TypeDescKind.UNION) {
+                BallerinaUnionTypeSymbol ballerinaUnionTypeSymbol = (BallerinaUnionTypeSymbol) returnTypeSymbol;
```

Comment:
`BallerinaUnionTypeSymbol` is the implementation. Instead use the exposed interface  `UnionTypeSymbol`. Then we don't need to expose the impl module

Reply:
Thanks